src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -0,0 +1,157 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal abstract class JsonSeparatorNamingPolicy : JsonNamingPolicy
+    {
+        private readonly bool _lowercase;
+        private readonly char _separator;
+
+        internal JsonSeparatorNamingPolicy(bool lowercase, char separator) =>
+            (_lowercase, _separator) = (lowercase, separator);
+
+        public sealed override string ConvertName(string name)
+        {
+            // Rented buffer 20% longer that the input.
+            int bufferLength = (12 * name.Length) / 10;
+            char[]? buffer = bufferLength > JsonConstants.StackallocCharThreshold
+                ? ArrayPool<char>.Shared.Rent(bufferLength)
+                : null;
+
+            int resultLength = 0;
+            Span<char> result = buffer is null
+                ? stackalloc char[JsonConstants.StackallocCharThreshold]
+                : buffer;
+
+            void ExpandBuffer(ref Span<char> result)
+            {
+                var bufferNew = ArrayPool<char>.Shared.Rent(bufferLength *= 2);
+
+                result.CopyTo(bufferNew);
+
+                if (buffer is not null)
+                {
+                    ArrayPool<char>.Shared.Return(buffer, clearArray: true);
```

Comment:
You could avoid clearing the entire buffer length by calling `result.Clear()` instead.